### PR TITLE
Document Users API

### LINKED ISSUE #390: Document Users API
#### GET /users

Retrieves a list of users  based on provided filters. Returns a JSON response with an array of properly serialized users

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_" 
            "http://api.lvh.me:5000/users?filter[id]=1,2,3"
```
#### GET /user

Retrieves the current logged in user from the database. Returns a JSON object containing a properly serialized user.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_" 
            "http://api.lvh.me:5000/user"
```

**_Example Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "new.inital.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```
#### GET /users/email_available

Checks whether an email exists in the database and returns its availability.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_" 
            "http://api.lvh.me:5000/users/email_available?email=thesamuelralak@gmail.com"
```

**_Response when email is valid and exists_**

```
{
  "available": false,
  "valid": true
}
```

**_Response when email is valid and  doesn't exist_**

```
{
  "available": true,
  "valid": true
}
```

**_Response when email is invalid_**

```
{
  "available": true,
  "valid": false
}
```
#### GET /users/username_available

Checks whether a username is valid and exists in the database and returns its availability.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_"  
            "http://api.lvh.me:5000/users/username_available?username=samuelralak"
```

**_Response when username is valid and exists_**

```
{
  "available": false,
  "valid": true
}
```

**_Response when username is valid and  doesn't exist_**

```
{
  "available": true,
  "valid": true
}
```

**_Response when username is invalid_**

```
{
  "available": true,
  "valid": false
}
```
#### GET /users/:id

Retrieves a registered user from the database. Responds with a 404 status code and proper error messages if user is not found.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_"  
            "http://api.lvh.me:5000/users/2"
```

**_Example response when user exists_**

```
{
  "data": {
    "id": "2",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T19:36:37.180Z",
      "email": null,
      "name": null,
      "username": "codecorpsguy",
      "twitter": null,
      "biography": null,
      "website": null,
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": [
          {
            "id": "1",
            "type": "organizations"
          }
        ]
      },
      "organization_memberships": {
        "data": [
          {
            "id": "1",
            "type": "organization_memberships"
          }
        ]
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Response when user is not found_**

```
{
  "errors": [
    {
      "id": "RECORD_NOT_FOUND",
      "title": "Record not found",
      "detail": "Couldn't find User with 'id'=4",
      "status": 404
    }
  ]
}
```
#### POST /users/

Creates a new user and adds to the database if parametes are valid. It will respond with a 422 status code and proper error messages if a user is invalid.

**_Example Request - with email_**

```
curl -X POST -H "Accept: application/json" 
             -H "Content-Type: application/json" 
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "email": "thesamuelralak@gmail.com",
                        "username": "samuelralak",
                        "password": "samiralak1"
                      }
                    }
                  }' "http://api.lvh.me:5000/users"
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "4",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-05T01:49:33.544Z",
      "email": null,
      "name": null,
      "username": "samilralak",
      "twitter": null,
      "biography": null,
      "website": null,
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Example Invalid Response_**

```
{
  "errors": [
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/email"
      },
      "detail": "has already been taken",
      "status": 422
    },
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/username"
      },
      "detail": "has already been taken",
      "status": 422
    }
  ]
}
```
#### POST /users/forgot_password

Allows registered users to reset their password. Returns a user JSON object if email exists otherwise it will return a 422 status code with proper validation errors.

**_Example Request_**

```
curl -X POST -H "Accept: application/json" 
             -H "Content-Type: application/json" 
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "email": "thesamuelralak@gmail.com"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/forgot_password"
```

**_Example Valid Response_**

```
{
    "data": {
        "id": "3",
        "type": "users",
        "attributes": {
            "created_at": "2016-07-03T20:19:26.452Z",
            "email": "thesamuelralak@gmail.com",
            "name": null,
            "username": "samuelralak",
            "twitter": null,
            "biography": "editted",
            "website": "new.inital.com",
            "facebook_id": null,
            "facebook_access_token": null,
            "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
            "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
            "state": "signed_up",
            "theme": "light"
        },
        "relationships": {
            "categories": {
                "data": []
            },
            "user_categories": {
                "data": []
            },
            "organizations": {
                "data": []
            },
            "organization_memberships": {
                "data": []
            },
            "roles": {
                "data": []
            },
            "user_roles": {
                "data": []
            },
            "skills": {
                "data": []
            },
            "user_skills": {
                "data": []
            }
        }
    }
}
```

**_Example Invalid Response_**

```
{
  "errors": [
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/email"
      },
      "detail": "doesn't exist in the database",
      "status": 422
    }
  ]
}
```
#### POST /users/reset_password

Changes a registered user's password if token is valid otherwise it returns a JSON object containing proper error messages and a 422 status code

**_Example Request_**

```
curl -X POST -H "Accept: application/json" 
             -H "Content-Type: application/json" 
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "confirmation_token": "7d2c59211330e9add35e3fbefbaa8e31f14aaca2",
                        "password": "newpassword"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/reset_password"
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "new.inital.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Example Invalid Response_**

```
{
  "errors": [
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/password"
      },
      "detail": "couldn't be reset",
      "status": 422
    }
  ]
}
```
#### PATCH /users/:id

Allows admin to perform edits to a registered user.

**_Example Request_**

```
curl -X PATCH  -H "Accept: application/json" 
             -H "Content-Type: application/json"
             -H "Authorization: Bearer _insert-access-token-here_"  
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "email": "thesamuelralak.com",
                        "biography": "editted"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/3
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "new.inital.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Example Invalid Response - when user is Unauthorized_**

```
{
  "errors": [
    {
      "id": "ACCESS_DENIED",
      "title": "Access denied",
      "detail": "You are not authorized to perform this action on users.",
      "status": 401
    }
  ]
}
```
#### PATCH /users/me

Allows user to perform edits to the current logged in user. Return a JSON object containing a properly serialized user

**_Example Request_**

```
curl -X PATCH  -H "Accept: application/json" 
             -H "Content-Type: application/json"
             -H "Authorization: Bearer _insert-access-token-here_"  
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "website": test.com
                        "biography": "editted"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/me"
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "test.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by christopherstyles**
This update adds API documentation for Users. The following actions have been documented:
- Create a user – `POST /users`
- Find a user – `GET /users/{id}`
- Check email availability – `GET /users/email_available?email={email}`
- Check username availability – `GET
  /users/username_available?username={username}`
- Update a user – `PATCH /users/{id}`
- Update “me” – `PATCH /users/me`
- List users – `GET /users?filter[id]=1,2`
- Show authenticated user – `GET /user`

Resolves #390, closes #455.

cc: @JoshSmith

### CODE REVIEW

File: blueprint/api.apib [hunk 1]
```diff
@@ -210,38 +210,192 @@ Exchange credentials for `access_token`.
...
+### Update me [PATCH /users/me]
+
++ Request (application/vnd.api+json)
+
+    + Attributes
+        + data
+            + Include (User Resource Identifier)
+            + attributes (object)
+                + first_name: `Code` (string, optional)
+                + last_name: `Corps` (string, optional)
```
Comment (joshsmith) [inline blueprint/api.apib]: I would use something like `Jane` and `Doe` here, or `Test` and `User`, just to make it a little clearer what object I'm looking at.

File: blueprint/api.apib [hunk 2]
```diff
@@ -355,72 +532,88 @@ Exchange credentials for `access_token`.
...
++ organization_memberships (object)
+  + data(array[Organization Membership Resource Identifier])
++ roles (object)
+  + data(array[Role Resource Identifier])
++ user_roles (object)
+  + data(array[User Role Resource Identifier])
++ skills (object)
+  + data(array[Skill Resource Identifier])
++ user_skills (object)
+  + data(array[User Skill Resource Identifier])
```
Comment (joshsmith) [inline blueprint/api.apib:590]: I would order these completely alpha. I know your aim with grouping the related joins together but I think it reads easier to have it in alpha form.

File: blueprint/api.apib [hunk 3]
```diff
@@ -355,72 +532,88 @@ Exchange credentials for `access_token`.
...
-## User (object)
+## User Attributes (object)
++ created_at: `2016-07-08T03:03:51.967Z` (string)
 + email: `team@codecorps.org` (string)
++ first_name: `Code` (string)
++ last_name: `Corps` (string)
++ name: `Code Corps` (string)
++ username: `codecorps` (string)
++ twitter: `codecorps` (string)
++ biography: `Sample bio.` (string)
```
Comment (joshsmith) [inline blueprint/api.apib:565]: I would change the name, username, twitter to be more reflective of a person than an organization.

File: blueprint/api.apib [hunk 4]
```diff
@@ -210,38 +210,192 @@ Exchange credentials for `access_token`.
...
+    + Attributes
+        + data
+            + Include (User Resource Identifier)
+            + attributes (object)
+                + first_name: `Code` (string, optional)
+                + last_name: `Corps` (string, optional)
+                + website: `https://codecorps.org` (string, optional)
+                + biography: `Sample bio.` (string, optional)
+                + twitter: `codecorps` (string, optional)
+                + base64_photo_data: `data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7` (string, optional)
```
Comment (joshsmith) [inline blueprint/api.apib:306]: Same name comments here.

Is there no way to use the attributes from another object here, too?

File: blueprint/api.apib [hunk 5]
```diff
@@ -210,38 +210,192 @@ Exchange credentials for `access_token`.
...
+    + Attributes
+
+        + available: `true` (boolean)
+        + valid: `true` (boolean)
+
+### Check username availability [GET /users/username_available?username={username}]
+
++ Parameters
+
+    + username: codecorps (string, required)
```
Comment (joshsmith) [inline blueprint/api.apib]: Would change this to something like `testuser` maybe.

**Comment by joshsmith** [comment]
@christopherstyles looks really good. Made some comments of some minor things we should consider fixing, and asked a question above.

File: blueprint/api.apib (hunk 1 above)
Comment (christopherstyles) [inline blueprint/api.apib]: Updated in 82109d7.

File: blueprint/api.apib (hunk 2 above)
Comment (christopherstyles) [inline blueprint/api.apib:590]: Agreed, updated in 75a2baa.

File: blueprint/api.apib (hunk 3 above)
Comment (christopherstyles) [inline blueprint/api.apib:565]: Updated in 82109d7.

File: blueprint/api.apib (hunk 4 above)
Comment (christopherstyles) [inline blueprint/api.apib:306]: Also updated with 82109d7. 

Re: attributes from another object: we could create _another_ set of data structure objects for update params and create params, but I don’t think they’d really be reused that much if at all. The `User Attributes` object has attributes like `created_at`, `photo_large_url`, etc, that aren’t allowed in update/create actions.

File: blueprint/api.apib (hunk 5 above)
Comment (christopherstyles) [inline blueprint/api.apib]: Updated in 82109d7.

**Comment by joshsmith** [comment]
@christopherstyles this looks good to me. I would merge at will.

**Pull request merged**